File: frontend/lib/services/auth_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../config/config.dart';
import '../core/api_client.dart';
import '../core/api_exception.dart';
import '../models/login_response.dart';
import '../models/usuario.dart';
import '../utils/session_manager.dart';

class AuthService {
  const AuthService();

  Future<Usuario> registrar({
    required String correo,
    required String contrasena,
    required String nombre,
    required String telefono,
    required int idCarrera,
  }) async {
    final usuario = Usuario(
      idUsuario: 0,
      correo: correo,
      nombre: nombre,
      tel: telefono,
      idCarrera: idCarrera,
    );

    try {
      final response = await http
          .post(
            Uri.parse('${Config.apiUrl}/register'),
            headers: ApiClient.jsonHeaders(withAuth: false),
            body: jsonEncode(
              usuario.toRegisterJson(
                contrasena: contrasena,
                idCarrera: idCarrera,
              ),
            ),
          )
          .timeout(Config.apiTimeout);

      if (response.statusCode == 200) {
        return Usuario.fromJson(
          jsonDecode(response.body) as Map<String, dynamic>,
        );
      }
      throw ApiException.fromResponse(response.statusCode, response.body);
    } on ApiException {
      rethrow;
    } catch (_) {
      throw ApiException(
        'No se pudo conectar con el servidor (${Config.apiUrl})',
      );
    }
  }

  /// POST /login — HU-02 con JWT.
  Future<LoginResponse> login({
    required String correo,
    required String contrasena,
  }) async {
    try {
      final response = await http
          .post(
            Uri.parse('${Config.apiUrl}/login'),
            headers: ApiClient.jsonHeaders(withAuth: false),
            body: jsonEncode({
              'correo': correo.trim(),
              'contrasena': contrasena,
            }),
          )
          .timeout(Config.apiTimeout);

      if (response.statusCode == 200) {
        final login = LoginResponse.fromJson(
          jsonDecode(response.body) as Map<String, dynamic>,
        );
        await SessionManager.iniciarSesion(
          LoginSessionData(token: login.token, usuario: login.usuario),
        );
        return login;
      }
      throw ApiException.fromResponse(response.statusCode, response.body);
    } on ApiException {
      rethrow;
    } catch (_) {
      throw ApiException(
        'No se pudo conectar con el servidor (${Config.apiUrl})',
      );
    }
  }
}

File: web/src/features/auth/auth-service.ts
import { ApiException, authRequest, getApiBaseUrl } from "./http";
import { setSession } from "./session";
import {
  parseCarrera,
  parseLoginResponse,
  parseUsuario,
  type Carrera,
  type LoginPayload,
  type LoginResponse,
  type RegisterPayload,
  type Usuario,
} from "./types";

/** POST /register — mirrors Flutter `AuthService.registrar`. */
export async function register(payload: RegisterPayload): Promise<Usuario> {
  try {
    const json = await authRequest<unknown>("/register", {
      method: "POST",
      withAuth: false,
      body: {
        correo: payload.correo.trim(),
        contrasena: payload.contrasena,
        nombre: payload.nombre.trim(),
        tel: payload.tel.trim(),
        idCarrera: payload.idCarrera,
      },
    });
    return parseUsuario(json);
  } catch (err) {
    if (err instanceof ApiException) throw err;
    throw new ApiException(
      `No se pudo conectar con el servidor (${getApiBaseUrl()})`,
    );
  }
}

/** POST /login — mirrors Flutter `AuthService.login` + SessionManager.iniciarSesion. */
export async function login(payload: LoginPayload): Promise<LoginResponse> {
  try {
    const json = await authRequest<unknown>("/login", {
      method: "POST",
      withAuth: false,
      body: {
        correo: payload.correo.trim(),
        contrasena: payload.contrasena,
      },
    });
    const loginResponse = parseLoginResponse(json);
    if (!loginResponse.token) {
      throw new ApiException("Respuesta de login sin token");
    }
    setSession({
      token: loginResponse.token,
      usuario: loginResponse.usuario,
    });
    return loginResponse;
  } catch (err) {
    if (err instanceof ApiException) throw err;
    throw new ApiException(
      `No se pudo conectar con el servidor (${getApiBaseUrl()})`,
    );
  }
}

/** GET /carreras — used by registro (same as Flutter CatalogoService). */
export async function fetchCarreras(): Promise<Carrera[]> {
  try {
    const json = await authRequest<unknown>("/carreras", {
      method: "GET",
      withAuth: false,
    });
    if (!Array.isArray(json)) {
      throw new ApiException("Respuesta de carreras inválida");
    }
    return json.map(parseCarrera);
  } catch (err) {
    if (err instanceof ApiException) throw err;
    throw new ApiException(
      `No se pudo conectar con el servidor (${getApiBaseUrl()})`,
    );
  }
}
